File: CSharp/121-BestTimeToBuyAndSellStock/121BestTimeToBuyAndSellStock.cs
// SLIDING WINDOW
// Time Complexity: O(n)
//  We have to go through potentially all numbers in prices.
// Space Complexity: O(1) 
//  We are just traversing the passed in prices
public class Solution {
    public int MaxProfit(int[] prices) {
        int min = prices[0];
        int profit = 0;
        foreach(int today in prices){
            profit = Math.Max(profit, today - min);
            min = Math.Min(min, today);
        }
        return profit;
    }
}

File: Javascript/121-BestTimeToBuyAndSellStock/121BestTimeToBuyAndSellStock.js
/*

Time Complexity: O(n)
Must go through every element of prices.

Space Complexity: O(1)
no matter the size of prices you just need the buy and profit variables.

*/

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let buy = prices[0];
    let profit = 0;
    for (let i = 1; i < prices.length; i++){
        if (prices[i] < buy)
            buy = prices[i];
        else if (prices[i] - buy > profit) {
            profit = prices[i] - buy;
        }
    }
    return profit;
};